### PR TITLE
Feature video 4review

### PR DESCRIPTION
**Pull request by Ragaad**

### CODE REVIEW

File: src/Video.cc [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+Video::Video( const Video &video){
+
+  _video = new VideoData(*video._video);
+
+}
+
+ void Video::operator=(const Video &vid){
+
+     VideoData *temp = _video;
```
Comment (vishakha041) [inline src/Video.cc]: Don't need temp

Comment (luisremis) [inline src/Video.cc]: no need for extra variable.

File: src/Video.cc [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+  _video = new VideoData(buffer, size, path);
+}
+
+Video::Video( const Video &video){
+
+  _video = new VideoData(*video._video);
+
+}
+
+ void Video::operator=(const Video &vid){
```
Comment (vishakha041) [inline src/Video.cc]: Try to keep parameter names consistent. So line 69 you say Video &video and here it is Video &vid.

File: src/Video.cc [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+
+using namespace VCL;
+
+using namespace cv; // OpenCV namespace
+
+void Video::store(const std::string &video_id, Format video_format,
+            bool store_metadata)
+{
+  _video->write(video_id, video_format, store_metadata);
```
[the remark this answers is not in the document]
Comment (vishakha041) [reply src/Video.cc]: And fix spaces between functions.

File: src/Video.cc [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+  _video->perform_operations();
+}
+
+std::string Video::create_unique(const std::string &path,
+                Format format)
+{
+  _video->create_unique(path, format);
```
Comment (vishakha041) [inline src/Video.cc]: I think you can just return output of create_unique()

File: src/VideoData.h [hunk 5]
```diff
@@ -0,0 +1,718 @@
...
+    /*  *********************** */
+        /**
+         *  Extends Operation, reads Video from the file system
+         */
+        class Read : public Operation {
+        private:
+            /** The full path to the object to read */
+            std::string _fullpath;
+            int _start; // specify the starting reading point of the video
+            int _stop;  //specify the ending point of the reading it represnet sa frame index
```
Comment (luisremis) [inline src/VideoData.h]: remove these members.

Comment (vishakha041) [inline src/VideoData.h]: Start, stop, step can just make one Interval struct and maybe pass it as argument to operations?

File: src/VideoData.h [hunk 6]
```diff
@@ -0,0 +1,718 @@
...
+            virtual OperationType get_type() = 0;
+        };
+
+    /*  *********************** */
+    /*       READ OPERATION     */
+    /*  *********************** */
+        /**
+         *  Extends Operation, reads Video from the file system
+         */
+        class Read : public Operation {
```
Comment (luisremis) [inline src/VideoData.h]: let's make sure we need this. @luisremis will help you with that.

Comment (vishakha041) [inline src/VideoData.h]: Does this need to be an operator if all its going to do is open the file. Could you not open the file in constructor or before perform_operations(). In Image, there is more code for TDB stuff

File: test/unit_tests/VideoData_test.cc [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+
+
+TEST_F(VideoDataTest, DefaultConstructor)
+{
+    VCL::VideoData video_data;
+    long input_frame_count = video_data.get_frame_count();
+    long test_frame_count;
+    cv::VideoCapture testVideo = cv::VideoCapture();
+    test_frame_count= testVideo.get(CV_CAP_PROP_FRAME_COUNT);
+    std::cout<< input_frame_count <<"=="<< test_frame_count <<std::endl;
```
Comment (luisremis) [inline test/unit_tests/VideoData_test.cc]: remove prints

File: test/unit_tests/VideoData_test.cc [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+
+
+class VideoDataTest : public ::testing::Test {
+ protected:
+    virtual void SetUp() {
+        _video = "videos/Megamind.avi";
+        _inputVideo = cv::VideoCapture("videos/Megamind.avi");
+
+    }
+    void compare_mat_mat(cv::Mat &cv_img, cv::Mat &img)
```
Comment (luisremis) [inline test/unit_tests/VideoData_test.cc]: if this method is the same as in image, create a common file for both tests

File: include/Image.h [hunk 9]
```diff
@@ -67,6 +59,28 @@ namespace VCL {
...
+                    LZ4 = 3,
+                    BLOSC = 4,
+                    BLZ4 = 5,
+                    BLZ4HC =  6,
+                    BSNAPPY = 7,
+                    BZLIB = 8,
+                    BZSTD = 9,
+                    RLE = 10 };
+
+                    enum class Format{
```
Comment (luisremis) [inline include/Image.h]: fix indent.

File: include/utils.h [hunk 10]
```diff
@@ -54,6 +54,15 @@ namespace VCL {
                     BZSTD = 9,
                     RLE = 10, };
 
+    enum class Format : int { NONE = 0,
```
[the remark this answers is not in the document]
Comment (Ragaad) [reply include/utils.h]: Done

File: include/Image.h [hunk 11]
```diff
@@ -45,14 +45,6 @@
 namespace VCL {
     class ImageData;
 
-    /*  *********************** */
-    /*        IMAGEFORMAT       */
-    /*  *********************** */
-    /**
-     *  Determines what kind of image it is
-     */
-    enum ImageFormat { NONE, JPG, PNG, TDB, };
```
[the remark this answers is not in the document]
Comment (Ragaad) [reply include/Image.h]: Done

File: include/Video.h [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+
+#include <opencv2/core.hpp>
+#include <opencv2/imgcodecs.hpp>
+#include <opencv2/videoio.hpp>
+
+
+#include "Exception.h"
+#include "utils.h"
+
+//#include "VideoUtils.h"
```
[the remark this answers is not in the document]
Comment (Ragaad) [reply include/Video.h]: done

File: include/Video.h [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+    typedef cv::Rect Rectangle; // spcifiy an ROI inside a video
+
+    class Video {
+
+    private:
+
+    VideoData *_video; // Pointer to a VideoData object
+
+
+
```
[the remark this answers is not in the document]
Comment (Ragaad) [reply include/Video.h]: done

File: src/VideoData.h (hunk 5 above)
Comment (Ragaad) [reply src/VideoData.h]: removed and only the data members that set on the interval operator were used instead

File: src/VideoData.h (hunk 6 above)
Comment (Ragaad) [reply src/VideoData.h]: The read operation now is empty, so we can remove it if it is fine with everyone. I don't know if we need it.

**Comment by luisremis** [comment]
closing this pull request, needs rebase and fixes to avoid merging conflicts.

**Pull request closed without merging**